lib/puppet/network/http_pool.rb

```diff
@@ -61,28 +61,27 @@ def self.http_ssl_instance(host, port, verifier = Puppet::SSL::Validator.default
                             :verify => verifier)
   end
 
-  # Retrieve a connection for the given uri.
+  # Retrieve a connection for the given host and port.
   #
-  # @param uri [URI] The URI to connect to
+  # @param host [String] The host to connect to
+  # @param port [Integer] The port to connect to
+  # @param use_ssl [Boolean] Whether to use SSL, defaults to `true`.
   # @param ssl_context [Puppet::SSL:SSLContext, nil] The ssl context to use
-  #   when making HTTPS connections.
+  #   when making HTTPS connections. Required when `use_ssl` is `true`.
   # @return [Puppet::Network::HTTP::Connection]
   #
   # @api public
   #
-  def self.connection(uri, ssl_context: nil)
-    case uri.scheme
-    when 'https'
-      raise ArgumentError, "An ssl_context is required for HTTPS connections: #{uri}" unless ssl_context
+  def self.connection(host, port, use_ssl: true, ssl_context: nil)
+    if use_ssl
+      raise ArgumentError, "An ssl_context is required when connecting to 'https://#{host}:#{port}'" unless ssl_context
 
       verifier = Puppet::SSL::Verifier.new(ssl_context)
-      http_client_class.new(uri.host, uri.port, use_ssl: true, verifier: verifier)
-    when 'http'
-      Puppet.warning("An ssl_context is unnecessary for HTTP connections and will be ignored: #{uri}") if ssl_context
-
-      http_client_class.new(uri.host, uri.port, use_ssl: false)
+      http_client_class.new(host, port, use_ssl: true, verifier: verifier)
     else
-      raise ArgumentError, "Unsupported scheme '#{uri.scheme}'"
+      Puppet.warning("An ssl_context is unnecessary when connecting to 'http://#{host}:#{port}' and will be ignored") if ssl_context
+
+      http_client_class.new(host, port, use_ssl: false)
```

Comment:
why not pass `ssl_context` to the Connection.initialize method directly, and have our Connection implementation create the verifier. Also puppetserver registers its own http_client_class